### ISSUE TITLE
Track artist shows as "John Paul Young" for all tracks with multiple artists, for Subsonic (Navidrome) library.

### ISSUE THREAD

**Issue by cautious-bread** [music-assistant/hass-music-assistant#2965]
### What version of Music Assistant has the issue?

2.2.5

### What version of the Home Assistant Integration have you got installed?

2024.9.1

### Have you tried everything in the Troubleshooting FAQ and reviewed the Open and Closed Issues and Discussions to resolve this yourself?

- [X] Yes

### The problem

Browsing through my imported/synced Navidrome (Subsonic) library I noticed a LOT of tracks by "John Paul Young".
Digging around I noticed these are all tracks within Navidrome that have more that one contributing artist to the track and no hyperlink for the artist field.
My tracks are all meticulously tagged via Picard.
I'm not sure why it's picking this random artist or ignoring the info in tag (they display as they should in Navidrome and other media managers/players).

![image](https://github.com/user-attachments/assets/2896f6e9-9971-45fc-b0eb-e85163b3c44a)

![image](https://github.com/user-attachments/assets/fb6d0cd5-ebfc-4a56-9527-9971a6b49dd8)

### How to reproduce

Add Navidrome as a Music Provider via the Subsonic option.
Allow it to sync.
Browse through the artists and spot the anomaly.

### Music Providers

Subsonic (Navidrome).

### Player Providers

SONOS.

### Full log output

[music-assistant.log](https://github.com/user-attachments/files/17129601/music-assistant.log)

### Additional information

Running on a VM inside Proxmox.  Deployed via HA's automated method.

### What version of Home Assistant Core are your running

2024.9.3

### What type of installation are you running?

Home Assistant OS

### On what type of hardware are you running?

Alternative

**Comment by OzGav** [comment]
@Jc2k

**Comment by OzGav** [comment]
@khers

**Comment by Clusters** [comment]
I have the same issue for some time already, just with ZIGGY instead of John Paul Young.

**Comment by khers** [comment]
I can have a look but I suspect this is related to the way that Navidrome reports tracks with multiple artists. I will need to make sure the provider is doing the correct thing in these cases.

**Comment by kristianwh** [comment]
I have kind of the same problem, but with random artists appearing all over the place. Using Tidal and Jellyfin (but problem started before I added Jellyfin to MA). 
![Screenshot_20241025_103045_Home Assistant](https://github.com/user-attachments/assets/5c238a4a-ae2e-4df0-8b59-cdbc5ffe824f)

**Comment by OzGav** [comment]
@kristianwh inwoudl do a db reset and start it again to eliminate problems from many versions ago. MA SETTINGS>> MUSIC CONTROLLER >> ADVANCED

**Comment by kristianwh** [comment]
@OzGav thanks, I think that did the trick!

**Comment by OzGav** [comment]
@cautious-bread @Clusters Can you both try a db reset as well to eliminate a core MA issue

**Comment by cautious-bread** [comment]
I recently removed my Subsonic music source because of the above issue and instead went with the "File System (remote share)" option as it _appeared_ to read the artists correctly.

I did the database reset above anyway, as I'm having a different issue now:  Some of the track listings for artists show the incorrect album against the track (from another artist), which also causes these tracks to be played when playing the other album, even they they don't show in said album.

For example;
For Gorillaz I have the "Demon Days" and self-titled "Gorillaz" albums.
For The Defiant I have the "If We're Really Being Honest" album.

Under the artist "Gorillaz" MA shows only the "Demon Days" album, but ALL the tracks for Gorillaz, but the Demon Days tracks showing against the Honest album from The Defiant.

Yet, under The Defiant, it shows the one correct Honest album and tracks.  No hint of Gorillaz tracks.  But... if I play the Honest album, it will queue up tracks from Demon Days scattered in the track listing.

I understand this is not the same as the initial problem I reported above, I'm just feeding back some further information around the library/database in case it sparks anyone's thoughts around the problem.

![image](https://github.com/user-attachments/assets/ef266851-dfa0-484a-a929-273cb3f33eeb)

![image](https://github.com/user-attachments/assets/c9f6491c-85be-48f1-bf3f-5c154a8ec63e)

![image](https://github.com/user-attachments/assets/85248291-6a08-4e39-9495-e2f3700f7013)

**Comment by OzGav** [comment]
@cautious-bread if you properly tagged your music with Picard then this would be unusual. If they have been properly tagged then please open your own issue

**Comment by dyno0919** [comment]
I am having the same issue originally posted by cautious-bread. A db reset does not fix it. Although when I remove the subsonic provider and re add it, in the first few moments before it's fully synched, the correct artists show up, then shortly after they revert to the wrong artist (Idina Menzel in my case... Curse you frozen soundtrack...)

**Comment by cautious-bread** [comment]
I decided to remove the File Share music provider and go back to the Subsonic music provider to test this - clearing the database throughout.
Unfortunately it's still naused for me...

![image](https://github.com/user-attachments/assets/7f2844a4-d998-4ce2-a45b-10ca98130c14)

![image](https://github.com/user-attachments/assets/7a50e1ac-e118-43a5-85fa-9861f14a359e)

😞

**Comment by khers** [comment]
IIRC the unifying factor here is Navidrome. I don't use Navidrome and therefore have a difficult time testing against it. IMO the way they handle multi-artist tracks is the cause here. The long term solution will be to move to using the newer opensubsonic return types which handle this in a much more sane way, however to the best of my knowledge none of the endpoints available for interacting with an opensubsonic server support these new data types. For the short term I see two options:

1. Someone that wants Navidrome support (and can test against a Navidrome server regularly) can submit patches to fix (I am happy to test any such submissions against my Gonic server to ensure we don't regress there).
2. Anyone who wants Navidrome support can ask the devs there to re-asses their handling of multi-artist tracks (to more closely match what Subsonic the original did).

I test against Gonic regularly because it is the server I use, so I can guarantee that it works there.

**Comment by OzGav** [comment]
@khers should we add a caution in the docs about Navidrome?

**Comment by dyno0919** [comment]
So I'm unfortunately only a novice coder and I have no experience in git/github, but the issue appears to be in the parse_track function in sonic_provider.py. Navidrome only creates artists for album artists, so if a track has an artist that does not have its own corresponding album, that artist will not have an id. Starting on line 409, there's a comment which seems to acknowledge this issue, but I'm not sure if it's been addressed properly. Again, I'm definitely inexperienced, but from my reading through the code, it looks like UNKOWN_ARTIST_ID is never redefined for each track, it's initialized as "fake_artist_unknown", but that same string is used to set the artist id for every artist that an id can't be found for. That would explain why we're having the same single artist show up for all of these tracks...

**Comment by rfporter** [comment]
Same behavior here on MA server 2.4.0b2 and Navidrome 0.50.1
If the track artist does not have an album in the library, it is replaced with James Horner. 
Is gonic the preferred subsonic music server? It seems like Navidrome is more popular.

**Comment by Clusters** [comment]
> @cautious-bread @Clusters Can you both try a db reset as well to eliminate a core MA issue

For my issues, I can confirm, that your suggested Controller reset has done the trick. It also resolved some weird playlist bugs for me. Thanks a lot!!!

**Comment by khers** [comment]
@cautious-bread I use this for a web front end to Gonic: https://github.com/tamland/airsonic-refix

I am not certain that Gonic reports the disc number but I can double check that  (it could be that I haven't hit a multi-disc album in MA yet, going to go look at Physical Graffiti)

**Comment by khers** [comment]
> @khers should we add a caution in the docs about Navidrome?

Yeah, that is probably the correct thing to do. I can try and write up a doc update this evening.

**Comment by khers** [comment]
> So I'm unfortunately only a novice coder and I have no experience in git/github, but the issue appears to be in the parse_track function in sonic_provider.py. Navidrome only creates artists for album artists, so if a track has an artist that does not have its own corresponding album, that artist will not have an id. Starting on line 409, there's a comment which seems to acknowledge this issue, but I'm not sure if it's been addressed properly. Again, I'm definitely inexperienced, but from my reading through the code, it looks like UNKNOWN_ARTIST_ID is never redefined for each track, it's initialized as "fake_artist_unknown", but that same string is used to set the artist id for every artist that an id can't be found for. That would explain why we're having the same single artist show up for all of these tracks...
> 
> Edit: More specifically, I think line 420 is the issue. The unique id is correctly set on line 415, but then on line 420, UNKNOWN_ARTIST_ID is used again instead of the unique id.
> 
> Edit 2: I think my initial comments were somewhat wrong. UNKNOWN_ARTIST_ID probably isn't supposed to be changed. But I definitely think the line in my first edit is the issue.

@dyno0919 If you can try changing line 420 to the following:

```item_id=f"{NAVI_VARIOUS_PREFIX}{sonic_song.artist}",```
And then reset the db and rescan, it could be that when putting together the hack to work  around Navidrome having artists on songs without entries in their artist table that I missed changing the lookup ID in the mapping. A big thank you for doing the digging on this!

**Comment by khers** [comment]
I have made the suggested change in this PR: https://github.com/music-assistant/server/pull/1758 once that is merged (and a beta release cut from it) can a Navidrome user try it and see if that addresses the artist issue?

@cautious-bread I also pass the disc number on to MA in this PR so it should address the multi-disc sorting as well.

**Comment by cautious-bread** [comment]
@khers Nice one.  I'll happily test it with Navidrome and Gonic when the next beta shows up.

**Comment by MentalHolyGlobe** [comment]
@khers Navidrome runs a [public demo instance](https://www.navidrome.org/demo/) that you may find useful for debugging issues specific to it in the future.

Thanks for the fix, I will also test it when it's in beta!

**Comment by khers** [comment]
This should be fixed in 2.4 beta 4

**Comment by cautious-bread** [comment]
> This should be fixed in 2.4 beta 4

YES!  Get in!  Nice work, and thank you so much.

I can confirm my original issue/bug has been resolved.  I'll test the disc number / Gonic issue another day and raise that separately if it's still a problem.

How do we know when this has been incorporated into the "stable" version of MA?  I originally went Beta to use the latest Sonos integration but I still want to move back to Stable eventually.

![image](https://github.com/user-attachments/assets/a0fdfe0d-ea22-4604-ae24-dbb0a34d0da7)

**Comment by MentalHolyGlobe** [comment]
Can also confirm the fix is working for me.

**Comment by khers** [comment]
Seeing disc number may need a database reset and a rescan.

**Comment by cautious-bread** [comment]
Already reset the library when adding Navidrome back in as a provider.  Disc numbers all good.  Same goes for Gonic.
My artists are still messed up when using Filesystem as the provider, but that's something for later-me to look into.

Everything is looking great mu end.  Thanks again!
